### PR TITLE
[CMake] [Darwin] Don't build the SDK overlays by default on Apple platforms

### PR DESCRIPTION
**Pull request by ian-twilightcoder**
The Apple SDKs have been providing the Darwin overlay since macOS 10.14.4, iOS 12.2, et al. More recently the SDK version has diverged from the Swift version making them incompatible. Stop building the overlay from Swift. Once the SDK overlays aren't being built, the clang overlays need to be built in testing.

rdar://115192929

### CODE REVIEW

**Comment by ian-twilightcoder** [comment]
@swift-ci smoke test macOS

**Comment by ian-twilightcoder** [comment]
@swift-ci smoke test macOS

**Comment by ian-twilightcoder** [comment]
This isn't enough to allow Swift to build against the current macOS 15.0 beta, but it's the upstream translation of what was done for rdar://115192929 and I think we'll need this all.

**Comment by ian-twilightcoder** [comment]
@swift-ci smoke test Linux Windows

**Comment by ian-twilightcoder** [comment]
@swift-ci smoke test Linux

**Comment by ian-twilightcoder** [comment]
@swift-ci smoke test Windows

File: CMakeLists.txt [hunk 1]
```diff
@@ -244,9 +244,16 @@ option(SWIFT_BUILD_CLANG_OVERLAYS
...
+# The SDK overlay is provided by the SDK itself on Darwin platforms.
+if (SWIFT_HOST_VARIANT_SDK IN_LIST SWIFT_DARWIN_PLATFORMS)
+  option(SWIFT_BUILD_DYNAMIC_SDK_OVERLAY
+      "Build dynamic variants of the Swift SDK overlay"
+      FALSE)
+else()
+  option(SWIFT_BUILD_DYNAMIC_SDK_OVERLAY
+      "Build dynamic variants of the Swift SDK overlay"
+      TRUE)
+endif()
```
Comment (al45tair) [inline CMakeLists.txt:256]: We should do this in a separate PR, not least because this isn't actually sufficient to stop it from building the Darwin module; the issue is this in `stdlib/public`:
```
if(SWIFT_BUILD_SDK_OVERLAY OR SWIFT_BUILD_TEST_SUPPORT_MODULES)
  add_subdirectory(Platform)
endif()
```
As `stdlib/CMakeLists.txt` makes clear:
```
option(SWIFT_BUILD_TEST_SUPPORT_MODULES
       "Whether to build StdlibUnittest and other test support modules. Defaults to On when SWIFT_BUILD_SDK_OVERLAY is On, or when SWIFT_INCLUDE_TESTS is On."
       "${SWIFT_BUILD_TEST_SUPPORT_MODULES_default}")
```
So if `SWIFT_INCLUDE_TESTS` is enabled, which it is by default, then we still build the overlays.

File: stdlib/public/SwiftShims/swift/shims/LibcShims.h [hunk 2]
```diff
@@ -170,11 +170,16 @@ long double _stdlib_squareRootl(long double _self) {
 // Apple's math.h does not declare lgamma_r() etc by default, but they're
 // unconditionally exported by libsystem_m.dylib in all OS versions that
 // support Swift development; we simply need to provide declarations here.
-#if defined(__APPLE__)
+// In the macOS 15.0, iOS 18.0, et al SDKs, math.h unconditionally declares
+// lgamma_r() when building for Swift. Detect those SDKs by checking for a
+// header which was added in those versions. (Redeclaring the function
+// would cause an error where `lgamma_r` is ambiguous between the SDK
+// `_math.lgamma_r` and this `SwiftShims.lgamma_r`.)
+#if defined(__APPLE__) && !__has_include(<_modules/_math_h.h>)
```
Comment (al45tair) [inline stdlib/public/SwiftShims/swift/shims/LibcShims.h:178]: Oh, I like that idea. :-)

File: utils/build_swift/build_swift/driver_arguments.py
Comment (al45tair) [inline utils/build_swift/build_swift/driver_arguments.py:1]: See previous; I think this should be in a different PR.

File: utils/build_swift/tests/expected_options.py
Comment (al45tair) [inline utils/build_swift/tests/expected_options.py:1]: See previous; I think this should be in a different PR.

**Review by al45tair — COMMENTED** [review]
I wouldn't mix the attempt to disable the `Darwin` module with the `lgamma_r` fix. The latter we could merge pretty much straight away as a separate PR on its own. There are some additional issues with the former.

File: CMakeLists.txt (hunk 1 above)
Comment (al45tair) [reply CMakeLists.txt:256]: (There's also @kubamracek's `embedded-darwin` module to worry about…)

File: stdlib/public/SwiftShims/swift/shims/LibcShims.h (hunk 2 above)
Comment (ian-twilightcoder) [reply stdlib/public/SwiftShims/swift/shims/LibcShims.h:178]: It was @mikeash 's 😄

Comment (mikeash) [reply stdlib/public/SwiftShims/swift/shims/LibcShims.h:178]: Gee, don't say it in public.... 😃

Comment (ian-twilightcoder) [reply stdlib/public/SwiftShims/swift/shims/LibcShims.h:178]: 🤣

**Comment by ian-twilightcoder** [comment]
@swift-ci smoke test macOS

**Comment by ian-twilightcoder** [comment]
Alright, trying again, stole a few changes from https://github.com/swiftlang/swift/pull/75902 that it looks like I missed in the original Radar

**Comment by ian-twilightcoder** [comment]
@swift-ci smoke test macOS

**Comment by ian-twilightcoder** [comment]
> I wouldn't mix the attempt to disable the `Darwin` module with the `lgamma_r` fix. The latter we could merge pretty much straight away as a separate PR on its own. There are some additional issues with the former.

It was easier to split the `lgamma_r` fix out and keep this for the `Darwin` module.

**Comment by ian-twilightcoder** [comment]
Looks like we're going to need https://github.com/swiftlang/swift/pull/75914 before we can do this one.

**Comment by ian-twilightcoder** [comment]
https://github.com/swiftlang/swift/pull/75914
@swift-ci smoke test macOS

**Comment by ian-twilightcoder** [comment]
https://github.com/swiftlang/swift/pull/75914

@swift-ci smoke test macOS

**Comment by ian-twilightcoder** [comment]
https://github.com/swiftlang/swift/pull/75914

@swift-ci smoke test

**Comment by ian-twilightcoder** [comment]
https://github.com/swiftlang/swift/pull/75914

@swift-ci smoke test

**Comment by ian-twilightcoder** [comment]
@swift-ci smoke test

**Comment by ian-twilightcoder** [comment]
This is stacked on top of https://github.com/swiftlang/swift/pull/75914 and has to wait for that one (or whatever it gets split out to) before this can be rebased retested and merged.

**Comment by ian-twilightcoder** [comment]
@swift-ci smoke test

**Comment by ian-twilightcoder** [comment]
@swift-ci smoke test

**Comment by ian-twilightcoder** [comment]
@swift-ci smoke test

**Comment by ian-twilightcoder** [comment]
@swift-ci smoke test

**Comment by ian-twilightcoder** [comment]
@swift-ci smoke test

**Comment by ian-twilightcoder** [comment]
@swift-ci smoke test

**Comment by ian-twilightcoder** [comment]
@swift-ci smoke test

**Comment by ian-twilightcoder** [comment]
@swift-ci  smoke test windows

**Comment by ian-twilightcoder** [comment]
@swift-ci smoke test Windows

**Comment by ian-twilightcoder** [comment]
@swift-ci smoke test

**Comment by ian-twilightcoder** [comment]
@swift-ci smoke test macOS
